.github/workflows/publish.yaml

```diff
@@ -0,0 +1,104 @@
+name: Publish
+on:
+  push:
+    branches:
+      - 'main'
+      - 'v.*'
+    tags:
+      - '.*'
+
+jobs:
+  unit-tests:
+    runs-on: ubuntu-20.04
+    steps:
+      - name: Cancel Previous Runs
+        uses: styfle/cancel-workflow-action@0.9.1
+        with:
+          access_token: ${{ github.token }}
+
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "1.17.1"
+
+      - name: Install Dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install python3-pip arping ndisc6
+          sudo pip3 install invoke semver pyyaml
+          go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0
+
+      - name: Unit Tests
+        run: |
+          inv test
+
+      - name: Lint
+        run: |
+          inv checkpatch
+          inv lint -e host
+          inv verifylicense
+          inv gomodtidy
+          inv checkchanges --action="run inv gomodtidy"
+          inv generatemanifests
+          inv checkchanges --action="run inv generatemanifests"
+          inv generateapidocs
+          inv checkchanges --action="run inv generateapidocs"
+
+  build-and-push:
+    needs: unit-tests
+    runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: false
+      matrix:
+        image: [controller, speaker]
+        arch: [amd64]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - uses: actions/setup-go@v2
+        with:
+          go-version: "1.17.1"
+
+      - name: Install Dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install python3-pip arping ndisc6
+          sudo pip3 install invoke semver pyyaml
+          go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.7.0
+
+      - name: Set up docker buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Build ${{ matrix.image }} binary and image
+        run: |
+          inv build --binaries=${{ matrix.image }} --registry=quay.io --repo=metallb --tag=${{ github.ref_name }} --architectures=${{ matrix.arch }}
+
+      - name: Login to Quay.io container registry
+        uses: docker/login-action@v1
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAY_USER }}
+          password: ${{ secrets.QUAY_PASSWORD }}
+
+      - name: Extract docker metadata
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: quay.io/metallb/${{ matrix.image }}
+          tags: |
+            type=raw,value=${{ github.ref_name }}-${{ matrix.arch }}
+
+      - name: Build and push ${{ matrix.image }} image
```

Comment:
This won't work.
We either choose to build the images with this, or with `inv build`.
If we want to go this way, we need to rethink the dockerfiles and have the build as part of the dockerfile so we are in control.
Given the comment about multiarch, we should consider what's faster, but the goal must be to have the images built locally (and by CI) with the same process done here.

Reply:
> You know best, but gh build-push action allows you to push to multiple registries (quay and dockerhub here) in one action, docs [here](https://github.com/docker/build-push-action/blob/master/docs/advanced/push-multi-registries.md). Might make the job simpler. Can also build multi-arch if that's something yall are interested in doing so.

Thanks for the info, currently this gh action is made to use `tasks.py `to build and publish metallb images, will revisit to this when we've Dockerfile itself as self contained one.

Reply:
yes Federico, this would break, I didn't have a closer look at Dockerfile earlier, now changed it to use `inv push-multiarch `command for this purpose.

Reply:
I agree, I filed https://github.com/metallb/metallb/issues/1408 for the follow up.